datafusion/physical-plan/src/work_table.rs

```diff
@@ -196,11 +206,22 @@ impl ExecutionPlan for WorkTableExec {
             0,
             "WorkTableExec got an invalid partition {partition} (expected 0)"
         );
-        let batch = self.work_table.take()?;
+        let ReservedBatches {
+            mut batches,
+            reservation,
+        } = self.work_table.take()?;
+        if let Some(projection) = &self.projection {
+            // We apply the projection
+            // TODO: it would be better to apply it as soon as possible and not only here
+            // TODO: an aggressive projection makes the memory reservation smaller, even if we do not edit it
```

Comment:
Could you expand on this a bit? Where would "as soon as possible" be? I'm not familiar with what WorkTableExec is or how it is used 🥲

Reply:
`WorkTableExec` is used for recursive CTEs. They are evaluated by executing the same query plan multiple time. The output of the previous iteration is injected as the value of the "work table" in the next iteration.

Here "as soon as possible" means that when we execute the previous iteration we don't need to save all the columns of the output of this iteration to feed them into the "work table" of the next iteration but only the columns that are actually needed.